### PR TITLE
[Feature] Support Side Adapter Network

### PR DESCRIPTION
**Pull request by angiecao**
## Motivation
Support SAN for Open-Vocabulary Semantic Segmentation
Paper: [Side Adapter Network for Open-Vocabulary Semantic Segmentation](https://arxiv.org/abs/2302.12242)
official Code: [SAN](https://github.com/MendelXu/SAN)

## Modification
- Added the parameters of backbone vit for implementing the image encoder of CLIP.
- Added text encoder code.
- Added segmentor multimodel encoder-decoder code for open-vocabulary semantic segmentation.
- Added SideAdapterNetwork decode head code.
- Added config files for train and inference.
- Added tools for converting pretrained models.
- Added loss implementation for mask classification model, such as SAN, Maskformer and remove dependency on mmdetection.
- Added test units for text encoder, multimodel encoder-decoder, san decode head and hungarian_assigner.

### Convert Models
**pretrained SAN model**
The official pretrained model can be downloaded from [san_clip_vit_b_16.pth](https://huggingface.co/Mendel192/san/blob/main/san_vit_b_16.pth) and [san_clip_vit_large_14.pth](https://huggingface.co/Mendel192/san/blob/main/san_vit_large_14.pth).
Use tools/model_converters/san2mmseg.py to convert offcial model into mmseg style.
`python tools/model_converters/san2mmseg.py <MODEL_PATH> <OUTPUT_PATH>`

**pretrained CLIP model**
Use the CLIP model provided by openai to train SAN. The CLIP model can be download from [ViT-B-16.pt](https://openaipublic.azureedge.net/clip/models/5806e77cd80f8b59890b7e101eabd078d9fb84e6937f9e85e4ecb61988df416f/ViT-B-16.pt) and  [ViT-L-14-336px.pt](https://openaipublic.azureedge.net/clip/models/3035c92b350959924f9f00213499208652fc7ea050643e8b385c2dac08641f02/ViT-L-14-336px.pt).
Use tools/model_converters/clip2mmseg.py to convert model into mmseg style.
`python tools/model_converters/clip2mmseg.py <MODEL_PATH> <OUTPUT_PATH>`

### Inference
test san_vit-base-16 model on coco-stuff164k dataset
`python tools/test.py ./configs/san/san-vit-b16_coco-stuff164k-640x640.py <TRAINED_MODEL_PATH>`

### Train
test san_vit-base-16 model on coco-stuff164k dataset
`python tools/train.py ./configs/san/san-vit-b16_coco-stuff164k-640x640.py --cfg-options model.pretrained=<PRETRAINED_MODEL_PATH>`

### Train on COCO-Stuff164k
|                 |       | mIoU  | mAcc  | pAcc  |
| --------------- | ----- | ----- | ----- | ----- |
| san-vit-base16  | official  | 41.93 | 56.73 | 67.69 |
|                 | mmseg | 41.93 | 56.84 | 67.84 |
| san-vit-large14 | official  | 45.57 | 59.52 | 69.76 |
|                 | mmseg | 45.78 | 59.61 | 69.21 |

### Evaluate on Pascal Context
|                 |       | mIoU  | mAcc  | pAcc  |
| --------------- | ----- | ----- | ----- | ----- |
| san-vit-base16  | official  | 54.05 | 72.96 | 77.77 |
|                 | mmseg | 54.04 | 73.74 | 77.71 |
| san-vit-large14 | official  | 57.53 | 77.56 | 78.89 |
|                 | mmseg | 56.89 | 76.96 | 78.74 |

### Evaluate on Voc12Aug
|                 |       | mIoU  | mAcc  | pAcc  |
| --------------- | ----- | ----- | ----- | ----- |
| san-vit-base16  | official  | 93.86 | 96.61 | 97.11 |
|                 | mmseg | 94.58 | 97.01 | 97.38 |
| san-vit-large14 | official  | 95.17 | 97.61 | 97.63 |
|                 | mmseg | 95.58 | 97.75 | 97.79 |

### CODE REVIEW

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @angiecao,

We'd like to express our appreciation for your valuable contributions to the mmsegmentation. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: configs/san/san-vit-b16_coco-stuff164k-640x640.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# By default, models are trained on 8 GPUs with 2 images per GPU
+train_dataloader = dict(batch_size=2)
+val_dataloader = dict(batch_size=1, dataset=dict(pipeline=test_pipeline))
+test_dataloader = val_dataloader
+
+data_preprocessor = dict(
+    size=crop_size,
+    mean=[122.7709, 116.7460, 104.0937],
+    std=[68.5005, 66.6322, 70.3232],
+    size_divisor=32,
```
Comment (xiexinch) [inline configs/san/san-vit-b16_coco-stuff164k-640x640.py:23]: We can't set the `size` and `size_divisor` at the same time.

File: configs/san/san-vit-b16_voc12aug-640x640.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    dict(type='ResizeShortestEdge', scale=crop_size, max_size=2560),
+    dict(type='LoadAnnotations'),
+    dict(type='PackSegInputs')
+]
+
+# By default, models are trained on 8 GPUs with 2 images per GPU
+train_dataloader = dict(batch_size=2)
+val_dataloader = dict(
+    batch_size=1, dataset=dict(reduce_zero_label=True, pipeline=test_pipeline))
+test_dataloader = val_dataloader
```
Comment (xiexinch) [inline configs/san/san-vit-b16_voc12aug-640x640.py:19]:
```suggestion
metainfo = dict(
    classes=('aeroplane', 'bicycle', 'bird', 'boat',
             'bottle', 'bus', 'car', 'cat', 'chair', 'cow', 'diningtable',
             'dog', 'horse', 'motorbike', 'person', 'pottedplant', 'sheep',
             'sofa', 'train', 'tvmonitor'),
    palette=[[128, 0, 0], [0, 128, 0], [128, 128, 0],
             [0, 0, 128], [128, 0, 128], [0, 128, 128], [128, 128, 128],
             [64, 0, 0], [192, 0, 0], [64, 128, 0], [192, 128, 0],
             [64, 0, 128], [192, 0, 128], [64, 128, 128], [192, 128, 128],
             [0, 64, 0], [128, 64, 0], [0, 192, 0], [128, 192, 0],
             [0, 64, 128]])

test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='ResizeShortestEdge', scale=crop_size, max_size=2560),
    dict(type='LoadAnnotations'),
    dict(type='PackSegInputs')
]

# By default, models are trained on 8 GPUs with 2 images per GPU
train_dataloader = dict(batch_size=2)
val_dataloader = dict(
    batch_size=1,
    dataset=dict(
        metainfo=metainfo,
        pipeline=test_pipeline))
test_dataloader = val_dataloader
```

Comment (xiexinch) [reply configs/san/san-vit-b16_voc12aug-640x640.py:19]: We should remove the `background` category.

File: mmseg/models/decode_heads/san_head.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+            num_fcs: int = 2,
+            qkv_bias: bool = True,
+            out_dims: int = 512,
+            final_norm: bool = True,
+            act_cfg: dict = dict(type='GELU'),
+            norm_cfg: dict = dict(type='LN'),
+    ):
+        super().__init__()
+
+        self.sos_token_format = sos_token_format
```
Comment (xiexinch) [inline mmseg/models/decode_heads/san_head.py:305]: Might add an assertion to this `sos_token_format`

File: mmseg/models/decode_heads/san_head.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+                TransformerEncoderLayer(
+                    embed_dims=embed_dims,
+                    num_heads=num_heads,
+                    feedforward_channels=mlp_ratio * embed_dims,
+                    num_fcs=num_fcs,
+                    qkv_bias=qkv_bias,
+                    act_cfg=act_cfg,
+                    norm_cfg=norm_cfg,
+                    batch_first=False))
+        '''
```
Comment (xiexinch) [inline mmseg/models/decode_heads/san_head.py:329]: Might remote this comment.

**Comment by xiexinch** [comment]
We should solve the conflict with dev-1.x.

File: mmseg/models/utils/san_layers.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+    return x
+
+
+class LayerNorm(nn.Module):
+    """A LayerNorm variant, popularized by Transformers, that performs point-
+    wise mean and variance normalization over the channel dimension for inputs
+    that have shape (batch_size, channels, height, width).
+
+    https://github.com/facebookresearch/ConvNeXt/blob/d1fa8f6fef0a165b27399986cc2bdacc92777e40/models/convnext.py#L119  # noqa B950
+    """
```
[the remark this answers is not in the document]
Comment (angiecao) [reply mmseg/models/utils/san_layers.py:392]: Yes, this method can be replaced by using `nn.LayerNorm` and `torch.permute` twice. But I think adding this LayerNorm variant might be a better choice. 
This variant is popularized by Transformers. It supports that the input shape can be directly (B, C, H, W). The input and output of the vision transformer layers are also in this shape, so there is no need to constantly adjust inputs and outputs between network layers as when using `nn.LayerNorm`.
In order to distinguish it from `nn.LayerNorm`, I changed the name to `LayerNorm2d`.

File: mmseg/models/utils/san_layers.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+                                                       src_len)
+        return attn_output, attn_output_weights  # .sum(dim=1) / num_heads
+    else:
+        return attn_output, None
+
+
+def cross_attn_layer(self: TransformerEncoderLayer, x, mem, attn_bias):
+    """Implementation of transformer layer with cross attention
+    Args:
+        self (TransformerEncoderLayer): The Module of vision transformer layer.
```
Comment (xiexinch) [inline mmseg/models/utils/san_layers.py:371]: It is suggested that `self` be renamed something else.

File: mmseg/models/utils/san_layers.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+    weights with self-attention.
+
+    Args:
+        self: self-attention layer
+        query, key, value: map a query and a set of key-value pairs to
+        an output. See "Attention Is All You Need" for more details.
+        attn_mask: mask that prevents attention to certain positions.
+    """
+    return cross_attn_with_self_bias_func(
+        query,
```
Comment (xiexinch) [inline mmseg/models/utils/san_layers.py:27]: Can the two methods be combined into one? And then the naming of the `self` param is a bit misleading.

File: mmseg/models/decode_heads/san_head.py [hunk 8]
```diff
@@ -0,0 +1,732 @@
...
+                        embed_dims=embed_dims,
+                        num_heads=num_heads,
+                        batch_first=False,
+                        bias=qkv_bias),
+                    ffn_cfgs=dict(
+                        type='FFN',
+                        embed_dims=embed_dims,
+                        feedforward_channels=mlp_ratio * embed_dims,
+                        act_cfg=act_cfg),
+                    operation_order=('norm', 'self_attn', 'norm', 'ffn')))
```
Comment (xiexinch) [inline mmseg/models/decode_heads/san_head.py:352]: Could we use self.cross_attn to control whether to add 'cross_attn' to the operation_order?

Comment (angiecao) [reply mmseg/models/decode_heads/san_head.py:352]: The implementation of cross attention for SAN implementations differs from that in mmcv. It also contains the process of calculating self-attention for query.
Concatenate self-attention weight of query with cross attention weight.
```python
    self_weight = (q * q_k).sum(dim=-1, keepdim=True) 
    total_attn_output_weights = torch.cat([attn_output_weights, self_weight], dim=-1)
    total_attn_output_weights = F.softmax(total_attn_output_weights, dim=-1)
```

Add weighted query to the final attention output.
```python
    attn_output = ( attn_output + self_weight * q_v) 
```

**Review by xiexinch — APPROVED** [review]

**Comment by zsc1220** [comment]
How to support open vocabulary prompting inference ?

**Comment by angiecao** [comment]
> How to support open vocabulary prompting inference ?

1. If you want a custom set of category names, you can define `model.text_encoder.vocabulary` in the config file and set `model.text_encoder.dataset_name` to None
```
model = dict(
    text_encoder=dict(dataset_name=None, 
                      vocabulary=['classA', 'classB', 'classC']))
```
2. If you want to set templates of prompts, you can define a new list of templates in the file mmseg/utils/get_templates.py and change model.text_encoder.templates to your template name in the config file.
```
PREDEFINED_TEMPLATES = {
    'custom':[
        'a photo of a {}.',
        'This is a photo of a {}',
        'There is a {} in the scene',
    ],
}
```
```
model = dict(text_encoder=dict(templates='custom'))
```
The implementation of prompt engineering is in function **template_encode** in the file mmseg/models/text_encoder/clip_text_encoder.py

**Comment by zsc1220** [comment]
Wow，Thank you so much ! By the way, does CAT-Seg algorithm also use this method ? if not , How to support open vocabulary prompting inference in  CAT-Seg algorithm?

**Comment by zsc1220** [comment]
```python
import numpy
print()
```
